node/pkg/libp2p/helper/app.go

```diff
@@ -0,0 +1,116 @@
+	"github.com/rs/zerolog/log"
+)
+
+type App struct {
+	Host host.Host
+	Bus  *bus.MessageBus
+}
+
+func New(bus *bus.MessageBus, h host.Host) *App {
+	return &App{
+		Bus:  bus,
+		Host: h,
+	}
+}
+
+func (a *App) Run(ctx context.Context) error {
+	defer a.subscribe(ctx)
+
+	sub, err := a.Host.EventBus().Subscribe(new(event.EvtPeerConnectednessChanged))
+	if err != nil {
+		return fmt.Errorf("event subscription failed: %w", err)
+	}
+
+	a.subscribeLibp2pEvent(ctx, sub)
+	return nil
+}
+
+func (a *App) subscribe(ctx context.Context) {
+	log.Debug().Str("Player", "Libp2pHelper").Msg("subscribing to libp2pHelper topics")
+	channel := a.Bus.Subscribe(bus.LIBP2P)
+	go func() {
+		log.Debug().Str("Player", "Libp2pHelper").Msg("starting libp2p subscription goroutine")
+		for {
+			select {
+			case msg := <-channel:
+				log.Debug().
+					Str("Player", "Libp2pHelper").
+					Str("from", msg.From).
+					Str("to", msg.To).
+					Str("command", msg.Content.Command).
+					Msg("libp2p received bus message")
+				go a.handleMessage(ctx, msg)
+			case <-ctx.Done():
+				log.Debug().Str("Player", "Libp2pHelper").Msg("stopping libp2pHelper subscription goroutine")
+				return
+			}
+		}
+	}()
+}
+
+func (a *App) handleMessage(ctx context.Context, msg bus.Message) {
+	if msg.To != bus.LIBP2P {
+		log.Debug().Str("Player", "Libp2pHelper").Msg("message not for libp2pHelper")
+		return
+	}
+	if msg.From != bus.ADMIN {
+		bus.HandleMessageError(errorSentinel.ErrBusNonAdmin, msg, "libp2pHelper received message from non-admin")
+	}
+
+	switch msg.Content.Command {
+	case bus.GET_PEER_COUNT:
+		log.Debug().Str("Player", "Libp2pHelper").Msg("get peer count msg received")
+		peerCount := len(a.Host.Network().Peers())
+		msg.Response <- bus.MessageResponse{Success: true, Args: map[string]any{"Count": peerCount}}
+	case bus.LIBP2P_SYNC:
+		log.Debug().Str("Player", "Libp2pHelper").Msg("libp2p sync msg received")
+		err := setup.ConnectThroughBootApi(ctx, a.Host)
+		if err != nil {
+			bus.HandleMessageError(err, msg, "failed to sync through boot api")
+			return
+		}
+		msg.Response <- bus.MessageResponse{Success: true}
+	}
+}
+
+func (a *App) subscribeLibp2pEvent(ctx context.Context, sub event.Subscription) {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				sub.Close()
+				return
+			case e := <-sub.Out():
+				a.handleEvent(ctx, e)
+			}
+		}
+	}()
+}
+
+func (a *App) handleEvent(ctx context.Context, e interface{}) {
+	log.Info().Str("Player", "Libp2pHelper").Msg("Disconnect event catched, triggering resync")
+	evt := e.(event.EvtPeerConnectednessChanged)
+	if evt.Connectedness == network.NotConnected {
+		for i := 1; i < 4; i++ {
+			// do not attempt immediate resync, but give some time
+			time.Sleep(time.Duration(i) * time.Minute)
+			err := setup.ConnectThroughBootApi(ctx, a.Host)
+			if err != nil {
+				log.Error().Err(err).Str("Player", "Libp2pHelper").Msg("Error occurred on boot API sync")
+			}
+		}
+	}
+}
```

Comment:
- if `subscribeLibp2pEvent` and `handleEvent` are designed to handle only `disconnect event`, which seems to be the case based on line 104 log, i suggest renaming the functions to be more specific
 - seems like the `for` loop is designed to attempt reconnect 3 times regardless of the previous reconnect attempt outcome. Shouldn't the loop break if reconnect is successful
 - this is a very minor suggestion, pls feel free to ignore, but how about changing `for` loop range from 0 to 3. I had to calculate how many times the loop runs from 1 to 4 but 0 to 3 is a muscle memory 😄

Reply:
- yes, do you have any good naming suggestion?
- yes, I'll make an update
- sure, I just didn't want to have +1 in time.Sleep, but if you insist, it could be updated

Reply:
I'm also terrible with naming but here is what copilot suggested: `subscribeLibp2pDisconnectEvent` and `handleDisconnectEvent`. Looks okay to me 😃 

And no need to update the for loop range. I didn't notice there was a reason the range is designed that way

Reply:
for second feedback, I think it's going to be a bit tricky to implement.

the first reason is that the sync function doesn't try to reconnect the disconnected host but try to connect all host in boot api.

the second reason is if the disconnection happened and the disconnected host restarts, it'll have different host id, which it can't be told if newly connected host is the same one or not

this is kind of workaround, for now, I'll just reflect the first feedback

Reply:
sounds good!